core/invoker/src/main/resources/application.conf

```diff
@@ -155,6 +155,7 @@ whisk {
       #aka 'How long should a container sit idle until we kill it?'
       idle-container = 10 minutes
       pause-grace = 50 milliseconds
+      keeping-duration = 60 minutes
```

Comment:
Is this the same thing as the `idle-container` config? Or is the warm container removed even if it has received an activation in the last 60 minutes?

Reply:
For different configuration.
* idle-container, e.g.

  If a FunctionPullingContainerProxy actor already executed some activations, but didn't execute activation for some time(50.milliseconds), FunctionPullingContainerProxy would goto(Paused), then it will create a timer

  ```
  case _ -> Paused   => startSingleTimer(IdleTimeoutName, StateTimeout, idleTimeout)
  ```
  we can see, it would send `StateTimeout`, then, StateTimeout would be handled by below codes, it would judge this FunctionPullingContainerProxy whether in `warmedContainerKeepingCount`

  ```
  when(Paused) {
  ...
    case Event(StateTimeout, data: WarmData) =>
      (for {
        count <- getLiveContainerCount(data.invocationNamespace, data.action.fullyQualifiedName(false), data.revision)
        (warmedContainerKeepingCount, warmedContainerKeepingTimeout) <- getWarmedContainerLimit(
          data.invocationNamespace)
      } yield {
        logging.info(
          this,
          s"Live container count: ${count}, warmed container keeping count configuration: ${warmedContainerKeepingCount} in namespace: ${data.invocationNamespace}")
        if (count <= warmedContainerKeepingCount) {
          Keep(warmedContainerKeepingTimeout)
        } else {
          Remove
        }
      }).pipeTo(self)
      stay
  ...
  }
  ```
* keeping-duration
Then, if this FunctionPullingContainerProxy is not in `warmedContainerKeepingCount`, it would be removed.
otherwise, this FunctionPullingContainerProxy would keep more keeping-duration time.